helios-testing/src/test/java/com/spotify/helios/testing/DefaultUndeployerTest.java

```diff
@@ -0,0 +1,131 @@
+
+import com.spotify.helios.client.HeliosClient;
+import com.spotify.helios.common.descriptors.Job;
+import com.spotify.helios.common.descriptors.JobId;
+import com.spotify.helios.common.protocol.JobDeleteResponse;
+import com.spotify.helios.common.protocol.JobUndeployResponse;
+
+import org.hamcrest.Matchers;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentMatcher;
+
+import java.util.List;
+
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.argThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class DefaultUndeployerTest {
+
+  private HeliosClient client;
+
+  private static final String HOST1 = "HOST1";
+  private static final String HOST2 = "HOST2";
+  private static final List<String> HOSTS = ImmutableList.of(HOST1, HOST2);
+  private static final JobId JOB_ID = new JobId("foo", "1", "hash");
+  private static final Job JOB = Job.newBuilder().setName("foo").setVersion("1").build();
+
+  @Before
+  public void before() {
+    client = mock(HeliosClient.class);
+  }
+
+  @Test
+  public void testSuccess() throws Exception {
+    final ListenableFuture<JobUndeployResponse> undeployFuture1 = Futures
+        .immediateFuture(new JobUndeployResponse(JobUndeployResponse.Status.OK, HOST1, JOB_ID));
+    final ListenableFuture<JobUndeployResponse> undeployFuture2 = Futures
+        .immediateFuture(new JobUndeployResponse(JobUndeployResponse.Status.OK, HOST2, JOB_ID));
+
+    //noinspection unchecked
+    when(client.undeploy(jobIdWithNameAndVersion(JOB_ID), anyString())).thenReturn(
+        undeployFuture1, undeployFuture2);
+
+    when(client.deleteJob(jobIdWithNameAndVersion(JOB_ID))).thenReturn(Futures.immediateFuture(
+        new JobDeleteResponse(JobDeleteResponse.Status.OK)));
+
+    final List<AssertionError> errors = new DefaultUndeployer(client).undeploy(JOB, HOSTS);
+
+    assertThat(errors, Matchers.empty());
+  }
+
+  @Test
+  public void testUndeployError() throws Exception {
+    final ListenableFuture<JobUndeployResponse> undeployFuture1 = Futures.immediateFuture(
+        new JobUndeployResponse(JobUndeployResponse.Status.FORBIDDEN, HOST1, JOB_ID));
+    final ListenableFuture<JobUndeployResponse> undeployFuture2 = Futures.immediateFuture(
+        new JobUndeployResponse(JobUndeployResponse.Status.HOST_NOT_FOUND, HOST2, JOB_ID));
+
+    //noinspection unchecked
+    when(client.undeploy(jobIdWithNameAndVersion(JOB_ID), anyString())).thenReturn(
+        undeployFuture1, undeployFuture2);
+
+    when(client.deleteJob(jobIdWithNameAndVersion(JOB_ID))).thenReturn(Futures.immediateFuture(
+        new JobDeleteResponse(JobDeleteResponse.Status.OK)));
+
+    final List<AssertionError> errors = new DefaultUndeployer(client).undeploy(JOB, HOSTS);
+
+    assertThat(errors.size(), equalTo(2));
```

Comment:
minor but I find that `assertThat(errors, hasSize(2))` outputs more helpful info when it fails than this form, as it prints the whole collection and it's elements